src/nvim/eval/encode.c

```diff
@@ -315,6 +315,55 @@ int encode_read_from_list(ListReaderState *const state, char *const buf,
       ga_append(gap, ')'); \
     } while (0)
 
+#define TYPVAL_ENCODE_CONV_PARTIAL(pt) \
+    do { \
+        int i; \
+        ga_concat(gap, "function("); \
+        if (&pt->pt_name != NULL) { \
+          unsigned len; \
+          char_u *p, *r, *s; \
+          len = 3; \
+          len += (unsigned)STRLEN(pt->pt_name); \
+          for (p = pt->pt_name; *p != NUL; mb_ptr_adv(p)) { \
+            if (*p == '\'') { \
+              len++; \
+            } \
+          } \
+          s = r = xmalloc(len); \
+          if (r != NULL) { \
+            *r++ = '\''; \
+            for (p = pt->pt_name; *p != NUL; ) { \
+              if (*p == '\'') { \
+                *r++ = '\''; \
+              } \
+            MB_COPY_CHAR(p, r); \
+            } \
+            *r++ = '\''; \
+            *r++ = NUL; \
+          } \
+          ga_concat(gap, s); \
+          xfree(s); \
```

Comment:
Why? `s` exists to remember what to free, while `r` is being used to move through the allocated memory and populate it.